bambi/interpret/effects.py

```diff
@@ -1,890 +1,857 @@
-# pylint: disable=ungrouped-imports
-from dataclasses import dataclass, field
-import itertools
+from typing import Any, Callable, Mapping, Optional
 
 import arviz as az
 import numpy as np
 import pandas as pd
-from pandas.api.types import is_categorical_dtype, is_string_dtype
-import xarray as xr
+from arviz import InferenceData
+from pandas import DataFrame, Series
+from pandas.api.types import (
+    is_float_dtype,
+    is_integer_dtype,
+    is_numeric_dtype,
+)
+from xarray import DataArray
 
-from bambi.models import Model
-from bambi.interpret.create_data import create_differences_data, create_predictions_data
+from bambi import Model
 from bambi.interpret.utils import (
-    average_over,
-    ConditionalInfo,
-    enforce_dtypes,
+    aggregate,
+    create_plot_config,
+    get_model_covariates,
+    get_response_and_target,
     identity,
-    merge,
-    VariableInfo,
 )
-from bambi.utils import get_aliased_name
-
 
-SUPPORTED_SLOPES = ("dydx", "eyex")
-SUPPORTED_COMPARISONS = {
-    "diff": lambda x, y: x - y,
-    "ratio": lambda x, y: x / y,
-}
+from .helpers import compare, create_inference_data
+from .plots import plot
+from .types import (
+    Conditional,
+    ConditionalParam,
+    Contrast,
+    ContrastParam,
+    Values,
+    Variable,
+)
 
 
-@dataclass
-class ResponseInfo:
-    """Stores metadata about the response variable for indexing data in az.InferenceData,
-    computing uncertainty intervals, and creating the summary dataframe in
-    'PredictiveDifferences'
+def validate_values(
+    values: Values, var_name: str, target_dtype: np.dtype | None = None
+) -> Series:
+    """Validate and convert user provided values to a Pandas Series.
 
     Parameters
     ----------
-    name : str
-        The name of the response variable.
-    target : str
-        The target of the response variable such as 'mean' or 'sigma'. Defaults to 'mean'.
-    lower_bound : float
-        The percentile of the lower bound of the uncertainty interval. Defaults to 0.03.
-    upper_bound : float
-        The percentile of the upper bound of the uncertainty interval. Defaults to 0.97.
+    values : Values
+        User-provided values as a list, numpy array, or pandas Series.
+    var_name : str
+        Name of the variable for error messages and Series naming.
+    target_dtype : np.dtype or None
+        Target data type to convert values to. If None, uses the natural dtype.
+
+    Returns
+    -------
+    Series
+        A pandas Series containing the validated and converted values.
+
+    Raises
+    ------
+    TypeError
+        If values are not a list, ndarray, or Series, or if they cannot be converted
+        to the target dtype.
+    ValueError
+        If the values container is empty.
     """
+    match values:
+        # Values can be a list
+        case list() as lst:
+            # Validate all elements are int or float
+            if not all(isinstance(x, (int, float)) for x in lst):
```

Comment:
Why does this work with only numeric values? Is there anything separate that handles the validation for non-numeric values (e.g., levels of a categorical variable)?

Reply:
Categorical values were validated within the `parse_conditional` function. After seeing this comment, I moved out the validation of categorical and numerical (float and integers) into separate functions.

Now, `get_contrast` and `get_conditional` will dispatch function calls to the appropriate validation functions depending on the series data type.